### PR TITLE
Fix regression on gufunc serialization

### LINKED ISSUE #6821: Regression: guvectorized functions can no longer be pickled in numba 0.53
This is a showstopper regression for users of guvectorized functions on dask.distributed:

```python
import numba
import pickle


@numba.guvectorize(["f8,f8[:]"], "()->()")
def double(x, out):
    out[:] = x * 2


print(pickle.dumps(double))
```

On Linux x64:
```bash
$ conda create -n test52 -c conda-forge python=3.8 numba=0.52
$ conda activate test52
$ python demo.py
b'\x80\x04\x95\x1a\x00\x00\x00\x00\x00\x00\x00\x8c\x0b__mp_main__\x94\x8c\x06double\x94\x93\x94.'
$ conda create -n test53 -c conda-forge python=3.8 numba=0.53
$ conda activate test53
$ python demo.py
Traceback (most recent call last):
  File "demo.py", line 10, in <module>
    print(pickle.dumps(double))
_pickle.PicklingError: Can't pickle <ufunc 'double'>: it's not the same object as __main__.double
```

### PR DESCRIPTION
**Pull request by sklam**
Fixes #6821.

### CODE REVIEW

File: numba/np/ufunc/gufunc.py [hunk 1]
```diff
@@ -26,6 +28,30 @@ def __init__(self, py_func, signature, identity=None, cache=None,
...
+    def _reduce_states(self):
+        gb = self.gufunc_builder
+        dct = dict(
+            py_func=gb.py_func,
+            signature=gb.signature,
+            identity=self._identity,
+            cache=gb.cache,
+            is_dynamic=self._is_dynamic,
+            targetoptions=gb.targetoptions,
+            typesigs=gb._sigs,
```
Comment (stuartarchibald) [inline numba/np/ufunc/gufunc.py:40]: Think `self._frozen` should be in here, whilst it has no impact on dynamic gufunc, it is transmittable state.

File: numba/tests/npyufunc/test_gufunc.py [hunk 2]
```diff
@@ -281,5 +283,90 @@ class TestGUVectorizeScalarParallel(TestGUVectorizeScalar):
...
+        got = np.zeros(1)
+        double(0.5, out=expect)
+        cloned(0.5, out=got)
+        self.assertPreciseEqual(expect, got)
+        # array version
+        expect = np.zeros_like(arr)
+        got = np.zeros_like(arr)
+        double(arr, out=expect)
+        cloned(arr, out=got)
+        self.assertPreciseEqual(expect, got)
```
Comment (stuartarchibald) [inline numba/tests/npyufunc/test_gufunc.py:368]: Wonder if these tests ought to also check, `self._frozen`, `self.gufunc_builder._sigs`, `self._is_dynamic` and `self._identity`, just to make sure the state is carrying appropriately?

**Review by stuartarchibald — COMMENTED** [review]
Thanks for the patch, couple of comments to address else looks good.

File: numba/np/ufunc/gufunc.py (hunk 1 above)
Comment (sklam) [reply numba/np/ufunc/gufunc.py:40]: fixed

File: numba/tests/npyufunc/test_gufunc.py (hunk 2 above)
Comment (sklam) [reply numba/tests/npyufunc/test_gufunc.py:397]: These attributes are now tested explicitly

File: numba/tests/npyufunc/test_gufunc.py [hunk 3]
```diff
@@ -281,5 +283,119 @@ class TestGUVectorizeScalarParallel(TestGUVectorizeScalar):
...
+        # attributes carried over
+        self.assertEqual(cloned._frozen, double._frozen)
+        self.assertEqual(cloned.identity, double.identity)
+        self.assertEqual(cloned.is_dynamic, double.is_dynamic)
+        self.assertEqual(cloned.gufunc_builder._sigs,
+                         double.gufunc_builder._sigs)
+        # expected value of attributes
+        self.assertTrue(cloned._frozen)
+
+        cloned.disable_compile()
```
Comment (stuartarchibald) [inline numba/tests/npyufunc/test_gufunc.py:307]: I don't think this has any effect but is harmless.

**Review by stuartarchibald — APPROVED** [review]
Thanks for the patch and fixes, looks good.

**Pull request merged**